File: android/src/main/kotlin/africa/ejara/trustdart/coins/XRP.kt
import africa.ejara.trustdart.Coin
import com.google.protobuf.ByteString
import wallet.core.java.AnySigner
import africa.ejara.trustdart.Numeric
import africa.ejara.trustdart.utils.toInt
import africa.ejara.trustdart.utils.toLong
import wallet.core.jni.CoinType
import wallet.core.jni.HDWallet
import wallet.core.jni.proto.Ethereum.SigningOutput
import wallet.core.jni.proto.Ripple

class XRP: Coin<String, CoinType>("XRP", CoinType.XRP)  {

    override fun signTransaction(
        path: String,
        txData: Map<String, Any>,
        mnemonic: String,
        passphrase: String
    ): String? {
        val wallet = HDWallet(mnemonic, passphrase)
        val signingInput = Ripple.SigningInput.newBuilder()
        val operation = Ripple.OperationPayment.newBuilder()

        operation.apply {
            amount = txData["amount"]!!.toLong()
            destination = txData["receiverAddress"] as String
            if (txData["memo"] != null) {
                destinationTag = txData["memo"]!!.toLong()
            }
        }

        signingInput.apply {
            account = txData["senderAddress" ] as String
            fee = txData["fee"]!!.toLong()
            sequence = txData["sequence"]!!.toInt()
            lastLedgerSequence = txData["lastLedgerSequence"]!!.toInt()
            privateKey = ByteString.copyFrom(wallet.getKey(coinType, path).data())
            opPayment = operation.build()
        }

        val sign = AnySigner.sign(signingInput.build(), coinType, SigningOutput.parser())
        return Numeric.toHexString(sign.encoded.toByteArray())

    }

}

File: ios/trustdart/Sources/trustdart/coins/XRP.swift
//
//  XRP.swift
//  trustdart
//
//  Created by Jay on 3/11/25.
//
import Foundation
import WalletCore

class XRP: Coin  {
    init() {
        super.init(name: "XRP", coinType: .xrp)
    }
    
    override func signTransaction(path: String, txData: [String : Any], mnemonic: String, passphrase: String) -> String? {
        let privateKey = HDWallet(mnemonic: mnemonic, passphrase: passphrase)?.getKey(coin: self.coinType, derivationPath: path)
        
        let operation = RippleOperationPayment.with {
            $0.destination = txData["receiverAddress"] as! String
            $0.amount = txData["amount"] as! Int64
            if (txData["memo"] != nil) {
                $0.destinationTag = UInt32(txData["memo"] as! String)!
            }
        }
        let input = RippleSigningInput.with {
            $0.fee = txData["fee"] as! Int64
            $0.sequence = txData["sequence"] as! UInt32
            $0.lastLedgerSequence = txData["lastLedgerSequence"] as! UInt32
            $0.account = txData["senderAddress" ] as! String
            $0.privateKey = privateKey!.data
            $0.opPayment = operation
        }
        
        let output: RippleSigningOutput = AnySigner.sign(input: input, coin: .xrp)
        return output.encoded.hexString
    }
    
}
